# Advanced Pandas techniques to boost performance!

https://engineering.upside.com/a-beginners-guide-to-optimizing-pandas-code-for-speed-c09ef2c6a4d6

Pycon Presentation - https://www.youtube.com/watch?v=HN5d490_KKk

We'll test and compare different methods of applying functions to pandas dataframes.
The function we'll be testing is the Haversine Distance formular which takes the cordinates of two points (lattitude and longitude) and calculates the straight line distance

In [1]:
import numpy as np

# Define a basic Haversine distance formula
def haversine(lat1, lon1, lat2, lon2):
    MILES = 3959
    lat1, lon1, lat2, lon2 = map(np.deg2rad, [lat1, lon1, lat2, lon2])
    dlat = lat2 - lat1 
    dlon = lon2 - lon1 
    a = np.sin(dlat/2)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon/2)**2
    c = 2 * np.arcsin(np.sqrt(a)) 
    total_miles = MILES * c
    return total_miles

## Remember, everytime you loop in Pandas a kitten dies.
Not really, but it's you who'll be suffering and I'll show you why

# Let's loop though a dataframe of cordinates and return a series with these distences

In [2]:
# Define a function to manually loop over all rows and return a series of distances
def haversine_looping(df):
    distance_list = []
    for i in range(0, len(df)):
        d = haversine(40.671, -73.985, df.iloc[i]['latitude'], df.iloc[i]['longitude'])
        distance_list.append(d)
    return distance_list

In [3]:
import pandas as pd

file_name = "https://raw.githubusercontent.com/rajeevratan84/datascienceforbusiness/master/new_york_hotels.csv"
df = pd.read_csv(file_name, encoding = "ISO-8859-1")
df.head()

ean_hotel_id                                     name  \
0        269955       Hilton Garden Inn Albany/SUNY Area   
1        113431     Courtyard by Marriott Albany Thruway   
2        108151                    Radisson Hotel Albany   
3        254756  Hilton Garden Inn Albany Medical Center   
4        198232  CrestHill Suites SUNY University Albany   

                 address1    city state_province postal_code  latitude  \
0     1389 Washington Ave  Albany             NY       12206  42.68751   
1  1455 Washington Avenue  Albany             NY       12206  42.68971   
2             205 Wolf Rd  Albany             NY       12205  42.72410   
3     62 New Scotland Ave  Albany             NY       12208  42.65157   
4  1415 Washington Avenue  Albany             NY       12206  42.68873   

   longitude  star_rating  high_rate  low_rate  
0  -73.81643          3.0   154.0272  124.0216  
1  -73.82021          3.0   179.0100  134.0000  
2  -73.79822          3.0   134.1700   84.1600  
3  -73.77638          3.0   308.2807  228.4597  
4  -73.81854          3.0   169.3900   89.3900

In [5]:
# This is a built in magic ipython command that provides timing for code executed in the cell
%timeit df['distance'] = haversine_looping(df)


282 ms ± 18.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


## Looping with iterrows()

In [6]:
%%timeit

# Haversine applied on rows via iteration
haversine_series = []
for index, row in df.iterrows():
    haversine_series.append(haversine(40.671, -73.985, row['latitude'], row['longitude']))
df['distance'] = haversine_series

184 ms ± 10.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


## We got a 300% improvement! But can we get it any faster?

In [7]:
%%timeit

# Timing apply on the Haversine function
df['distance'] = df.apply(lambda row: haversine(40.671, -73.985, row['latitude'], row['longitude']), axis=1)

53.1 ms ± 2.42 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


### We went from 600ms to 60ms, a ten fold increase in speed!

In [8]:
#!pip install line_profiler
%load_ext line_profiler

# Haversine applied on rows with line profiler
%lprun -f haversine df.apply(lambda row: haversine(40.671, -73.985, row['latitude'], row['longitude']), axis=1)


Timer unit: 1e-09 s

Total time: 0.13246 s
File: /tmp/ipykernel_17464/2315215438.py
Function: haversine at line 4

Line #      Hits         Time  Per Hit   % Time  Line Contents
     4                                           def haversine(lat1, lon1, lat2, lon2):
     5      1631    6386620.0   3915.8      4.8      MILES = 3959
     6      1631   40955570.0  25110.7     30.9      lat1, lon1, lat2, lon2 = map(np.deg2rad, [lat1, lon1, lat2, lon2])
     7      1631    4278078.0   2623.0      3.2      dlat = lat2 - lat1 
     8      1631    2287335.0   1402.4      1.7      dlon = lon2 - lon1 
     9      1631   45944776.0  28169.7     34.7      a = np.sin(dlat/2)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon/2)**2
    10      1631   18685405.0  11456.4     14.1      c = 2 * np.arcsin(np.sqrt(a)) 
    11      1631    3201832.0   1963.1      2.4      total_miles = MILES * c
    12      1631   10720079.0   6572.7      8.1      return total_miles

### We can note two things from the output.
1. Line 9 our trigonometric calcuation accounts for the bulk of our processing time, understandably so.
2. Each line is being hit 1631 times, that is because we're still **looping** so we iterate through every row in our dataframe

# Now for Vectorization!

- We've only been passing scalars to our Haversine function. Fortunately, all of the functions being used within the Haversine function, are also able to operate on arrays. 
- This makes the process of vectorizing our distance function quite simple as we can pass the entire series (columns) to it, performing all the calculations on the entire array simultaneously.

![alt text](https://github.com/rajeevratan84/datascienceforbusiness/blob/master/vectorization.png?raw=true)

In [9]:
%%timeit 

# Vectorized implementation of Haversine applied on Pandas series
df['distance'] = haversine(40.671, -73.985, df['latitude'], df['longitude'])

2.79 ms ± 224 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [10]:
# Run our line profiler to inspect further
%lprun -f haversine df['distance'] = haversine(40.671, -73.985, df['latitude'], df['longitude'])

Timer unit: 1e-09 s

Total time: 0.0143024 s
File: /tmp/ipykernel_17464/2315215438.py
Function: haversine at line 4

Line #      Hits         Time  Per Hit   % Time  Line Contents
     4                                           def haversine(lat1, lon1, lat2, lon2):
     5         1       1603.0   1603.0      0.0      MILES = 3959
     6         1     822626.0 822626.0      5.8      lat1, lon1, lat2, lon2 = map(np.deg2rad, [lat1, lon1, lat2, lon2])
     7         1     423559.0 423559.0      3.0      dlat = lat2 - lat1 
     8         1     367450.0 367450.0      2.6      dlon = lon2 - lon1 
     9         1   10218135.0 1.02e+07     71.4      a = np.sin(dlat/2)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon/2)**2
    10         1    2055353.0 2.06e+06     14.4      c = 2 * np.arcsin(np.sqrt(a)) 
    11         1     411943.0 411943.0      2.9      total_miles = MILES * c
    12         1       1779.0   1779.0      0.0      return total_miles

We can see that each line is hit once! exactly what we wanted, however, line 9 still accounts for the bulk of our proessing time. Can that be sped up? 

## Let's use numpy vetorization instead of pandas

In [11]:
%%timeit

# Vectorized implementation of Haversine applied on NumPy arrays (note we use .values to access the numpy series)
df['distance'] = haversine(40.671, -73.985, df['latitude'].values, df['longitude'].values)

372 μs ± 27.2 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


The slowest run took 8.09 times longer than the fastest. This could mean that an intermediate result is being cached.
1000 loops, best of 3: 274 µs per loop

### So we went from 609 ms using regular for loops, to 275 µs (or 0.275ms)!!! A 2000 fold increase in performance!